tsl/src/compression/compression.c

```diff
@@ -743,42 +736,41 @@ run_analyze_on_chunk(Oid chunk_relid)
 	ExecVacuum(NULL, &vs, true);
 }
 
-/* Find segment by index for setting the correct sequence number if
- * we are trying to roll up chunks while compressing
+/* Find segment by index index in relInfo for setting the correct sequence number if
+ * we are trying to roll up chunks while compressing.
+ *
+ * Returns -1 if none matching.
  */
-static Oid
-get_compressed_chunk_index(Relation compressed_chunk, int16 *uncompressed_col_to_compressed_col,
+static int8
+get_compressed_chunk_index(ResultRelInfo *resultRelInfo, int16 *uncompressed_col_to_compressed_col,
 						   PerColumn *per_column, int n_input_columns)
 {
-	ListCell *lc;
-	int i;
+	// Number of indexes could potentially be over max(int8).
+	// This would be highly unlikely and impractical.
+	Assert(resultRelInfo->ri_NumIndices < 128);
 
-	List *index_oids = RelationGetIndexList(compressed_chunk);
-
-	foreach (lc, index_oids)
+	for (int i = 0; i < resultRelInfo->ri_NumIndices; i++)
 	{
-		Oid index_oid = lfirst_oid(lc);
 		bool matches = true;
 		int num_segmentby_columns = 0;
-		Relation index_rel = index_open(index_oid, AccessShareLock);
-		IndexInfo *index_info = BuildIndexInfo(index_rel);
+		IndexInfo *index_info = resultRelInfo->ri_IndexRelationInfo[i];
 
-		for (i = 0; i < n_input_columns; i++)
+		for (int j = 0; j < n_input_columns; j++)
 		{
-			if (per_column[i].segmentby_column_index < 1)
+			if (per_column[j].segmentby_column_index < 1)
 				continue;
 
 			/* Last member of the index must be the sequence number column. */
-			if (per_column[i].segmentby_column_index >= index_rel->rd_att->natts)
+			if (per_column[j].segmentby_column_index >= index_info->ii_NumIndexAttrs)
 			{
 				matches = false;
 				break;
 			}
 
-			int index_att_offset = AttrNumberGetAttrOffset(per_column[i].segmentby_column_index);
+			int index_att_offset = AttrNumberGetAttrOffset(per_column[j].segmentby_column_index);
 
 			if (index_info->ii_IndexAttrNumbers[index_att_offset] !=
-				AttrOffsetGetAttrNumber(uncompressed_col_to_compressed_col[i]))
+				AttrOffsetGetAttrNumber(uncompressed_col_to_compressed_col[j]))
```

Comment:
in case of a table with `segmentby=s1,s2`
I think this method is not preapred to consider an index only on the `s2` column - is that true?

Reply:
thank you for answering - I was thinking that this is some general method - but now I see that this is only used to get the `sequence_num` for the group...

I wonder why the `SEQUENCE_NUM_GAP`  is there; lookup is done to get the seq number  - is the address space inside the GAPs are actually used?